framework/src/media/MediaPlayerImpl.cpp

```diff
@@ -544,16 +544,46 @@ void MediaPlayerImpl::playback()
 	ssize_t num_read = mInputDataSource->read(mBuffer,(int)mBufSize);
 	meddbg("num_read : %d\n", num_read);
 	if (num_read > 0) {
-		start_audio_stream_out(mBuffer, get_output_bytes_to_frame((unsigned int)num_read));
+		int ret = start_audio_stream_out(mBuffer, get_output_bytes_to_frame((unsigned int)num_read));
+		if (ret < 0) {
+			PlayerWorker& mpw = PlayerWorker::getWorker();
+			switch (ret) {
+			case AUDIO_MANAGER_XRUN_STATE:
+				meddbg("AUDIO_MANAGER_XRUN_STATE\n");
+				break;
+			default:
+				meddbg("audio manager error : %d\n", ret);
+				notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+				mpw.enQueue(&MediaPlayerImpl::stopByError, shared_from_this());;
+				break;
+			}
+		}
 	} else if (num_read == 0) {
 		notifyObserver(PLAYER_OBSERVER_COMMAND_FINISHIED);
 		stop();
 	} else {
+		PlayerWorker& mpw = PlayerWorker::getWorker();
+		meddbg("InputDatasource read error\n");
 		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
-		stop();
+		mpw.enQueue(&MediaPlayerImpl::stopByError, shared_from_this());;
 	}
 }
 
+void MediaPlayerImpl::stopByError()
```

Comment:
What is the difference between stopByError() and stopPlayer()?
It seems to the same of the internal logic.

Reply:
this is stopPlayer code.
```
	audio_manager_result_t result = stop_audio_stream_out();
	if (result != AUDIO_MANAGER_SUCCESS) {
		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
		meddbg("stop_audio_stream_out failed ret : %d\n", result);
		return;
	}

	mpw.setPlayer(nullptr);
	mCurState = PLAYER_STATE_READY;
```

if `stop_audio_stream_out` failed, mCurState is not changed. So playback is not stopped.
This makes massive error callback.